build: add --enable-isystem and change --enable-werror to enable -Werror

vasild opened the pull request:
Before this patch `./configure --enable-werror` would selectively
turn only some warnings into errors, not all, by using
`-Werror=foo -Werror=bar` instead of a plain `-Werror` which turns all 
warnings into errors.

To accommodate the new stricter build config, introduce 
`--enable-isystem` to optionally suppress warnings from external headers 
(e.g. boost, qt). That may be useful on systems that have them installed 
outside of `/usr/include`. On Linux the external headers are typically 
installed in `/usr/include` and so warnings from them are always suppressed.

vasild committed:
build: change --enable-werror to enable -Werror

Before this patch `./configure --enable-werror` would selectively
turn only some warnings into errors, not all, by using
`-Werror=foo -Werror=bar` instead of a plain `-Werror` which turns all
warnings into errors.

To accommodate the new stricter build config, use --enable-isystem to
suppress warnings from external headers (e.g. boost, qt) on systems
that have them installed outside of /usr/include. On Linux the external
headers are typically installed in /usr/include and so warnings from
them are suppressed by default.

vasild commented:
This works on FreeBSD 12 with Clang 9. Lets see if some other platform will break when using the enhanced `--enable-werror`.

hebasto commented:
> Suppress any warnings from Boost headers by marking them as system
> headers.

Suppressing _any_ warnings from Boost does not look great. Some related discussions about warnings from leveldb could be find in #16710 and #16722.

vasild commented:
@hebasto yes, I agree suppressing warnings from external headers (e.g. boost) is not good. However, it is arguable whether not "Werror"ing our own warnings in order to see warnings from external headers is justified. It would be most flexible to be able to choose what to do, depending on `./configure` options. I imagine something like this, maybe:

./configure options | behavior
---|---
--enable-werror | Full `-Werror`, any warnings are turned into errors. Most likely will fail the build due to warnings from external headers
--suppress-warnings-from-external-headers | Warnings from external headers are suppressed. Our own warnings are just shown (do not cause build failure).
--enable-werror --suppress-warnings-from-external-headers | Our warnings fail the build. External ones are not shown. Can be used to make sure our code is warning-free.

Empact commented on configure.ac:
```diff
@@ -1133,6 +1135,10 @@ dnl counter implementations. In 1.63 and later the std::atomic approach is defau
 BOOST_CPPFLAGS="-DBOOST_SP_USE_STD_ATOMIC -DBOOST_AC_USE_STD_ATOMIC $BOOST_CPPFLAGS"
 
+if test "x$enable_suppress_external_warnings" = "xyes"; then
+  BOOST_CPPFLAGS="`echo $BOOST_CPPFLAGS |sed 's/-I/-idirafter /g'`"
```
Why not isystem? It would more closely match the behavior of the -I it replaces.
https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html

vasild replied:
Because `-idirafter /usr/include` does not break gcc's `#include_next`, so I do not have to explicitly handle `/usr/include` in a different way.

Anyway, the `--enable-isystem` introduced in https://github.com/bitcoin/bitcoin/commit/572db3ce8103940c58cc90f4f770e989cf789835 in https://github.com/bitcoin/bitcoin/pull/14920 is superior to the above.

vasild replied:
I just rebased on top of https://github.com/bitcoin/bitcoin/commit/572db3ce8103940c58cc90f4f770e989cf789835 and ditched the above piece.

Empact replied:
Here's a much pared-down version of 572db3c that looks to be effective (applied to Boost only):
https://github.com/bitcoin/bitcoin/compare/master...Empact:2020-02-werror?expand=1

vasild replied:
https://github.com/bitcoin/bitcoin/compare/master...Empact:2020-02-werror looks good. I confirm it works as expected on FreeBSD 12 with clang 9 - it makes it possible to compile Bitcoin Core with full `-Werror`! \o/

Are you going to create PR from it? If yes, then I will ditch this PR.

nit: the new option's description `enable isystem includes for dependencies and stricter warning checks (default is no)` warrants `s/ and stricter warning checks//`

Thanks!

Empact replied:
Feel free to cherry-pick and modify the subject. Deferring to you as the original author unless you want to punt.

vasild replied:
Done. I also picked up the QT changes to get this working also with Clang 10.

Empact committed:
build: Optionally include dependency headers with -isystem

When configured with --enable-isystem.

Was necessary to split QT_INCLUDES into QT_INCLUDES and
QT_MOC_INCLUDES because moc does not understand -isystem, e.g.:

    Unknown options: isystem/usr/local/Cellar/qt/5.10.0_1/include/QtNetwork[...]

This patch treats just Boost and QT headers. The other external headers could be
added as well, should it be necessary.

Note -isystem should not be applied to /usr/include, see BITCOIN_SYSTEM_INCLUDE
for a helper to convert -I to -isystem with /usr/include excepted.

vasild commented:
@hebasto this PR went through some significant changes, I edited my comments accordingly and changed its status from "Draft" to "Open". Your re-review would be welcome.

vasild commented:
Closing this as it contains two not directly related changes - optional suppression of warnings from external headers and -Werror expansion. I will submit them separately.

vasild commented:
Took the idea of the first commit and re-implemented it in a (hopefully) better/shorter way in https://github.com/bitcoin/bitcoin/pull/18750
